packages/server/src/main/java/software/uncharted/terarium/hmiserver/aspects/HasProjectAccessAspect.java

```diff
@@ -0,0 +1,120 @@
+package software.uncharted.terarium.hmiserver.aspects;
+
+import java.lang.reflect.Method;
+import java.util.Optional;
+import java.util.UUID;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.annotation.Around;
+import org.aspectj.lang.annotation.Aspect;
+import org.aspectj.lang.reflect.MethodSignature;
+import org.springframework.expression.Expression;
+import org.springframework.expression.ExpressionParser;
+import org.springframework.expression.spel.standard.SpelExpressionParser;
+import org.springframework.expression.spel.support.StandardEvaluationContext;
+import org.springframework.http.HttpStatus;
+import org.springframework.security.access.AccessDeniedException;
+import org.springframework.stereotype.Component;
+import org.springframework.web.server.ResponseStatusException;
+import software.uncharted.terarium.hmiserver.annotations.HasProjectAccess;
+import software.uncharted.terarium.hmiserver.models.User;
+import software.uncharted.terarium.hmiserver.models.dataservice.project.Project;
+import software.uncharted.terarium.hmiserver.service.CurrentUserService;
+import software.uncharted.terarium.hmiserver.service.data.ProjectService;
+
+@Aspect
+@Component
+@Slf4j
+@RequiredArgsConstructor
+public class HasProjectAccessAspect {
+
+	private final ProjectService projectService;
+	private final CurrentUserService currentUserService;
+
+	private static final ExpressionParser expressionParser = new SpelExpressionParser();
+
+	/**
+	 * This aspect will be called around all methods of classes that have the HasProjectAccess annotation.
+	 * If the method itself is annotated, we'll let the method annotation take precedence.
+	 * @param joinPoint   The join point
+	 * @param annotation  The {@link HasProjectAccess} annotation
+	 * @return            The proceeding join point if the user has access, otherwise an exception will be thrown
+	 * @throws Throwable  If the user does not have access
+	 */
+	@Around("@within(annotation)")
```

Comment:
Out of curiosity, is this a compile time thing? or is this a runtime thing with reflections? 
Cause we have some dumb hibernate fetches that might not work well if it is the latter case.

Reply:
In this case its the later with `@Retention(RetentionPolicy.RUNTIME)` defined in the annotation definition.  One way to find out if its a problem! Those specific calls may need to be handled seperately